### ISSUE TITLE
Allow on-demand houskeeping for Prometheus scraping

### ISSUE THREAD

**Issue by bobrik** [google/cadvisor#1989]
We use cAdvisor with Prometheus. In this scenario you usually only care about metrics at exact moment of scraping and don't bother with intermediate results.

If you scrape your metrics every 60s, cAdvisor still pulls metrics from `sysfs` and whatnot every 1s, that's 60x overhead in the worst case (dynamic housekeeping may help a bit). With increased houskeeping interval we run into issues when prometheus scrapes can be misaligned with housekeeping and produce sawtooth.

For example, if your housekeeping interval is 30s, you can end up with either 1 or 2 updates per scrape depending on your luck. Something I found in the old tickets: `one` and `two` are containers that constantly burn a single CPU each, `USER` is from `node_exporter`.

![image](https://user-images.githubusercontent.com/89186/42249692-8fa290b6-7ee1-11e8-9dea-285662e6f40a.png)

I suggest to add a mode where cAdvisor only does on-demand housekeeping when asked for Prometheus metrics. This way we can lose short-lived containers, but that's something we can accept. For machines running lots of long-lived containers this should provide substantial efficiency benefits.

cc @dqminh

**Comment by bobrik** [comment]
To give some perspective: we have `cadvisor` instances burning more than one full CPU core with `tcp` and `udp` metrics disabled.

![image](https://user-images.githubusercontent.com/89186/42249856-5a5ec766-7ee2-11e8-8077-17ed26de9290.png)

**Comment by bobrik** [comment]
To illustrate difference in resource usage, let's consider the following ugly patch:

```diff
diff --git a/manager/container.go b/manager/container.go
index 295479f0..82993ae2 100644
--- a/manager/container.go
+++ b/manager/container.go
@@ -103,6 +103,10 @@ func jitter(duration time.Duration, maxFactor float64) time.Duration {
 }
 
 func (c *containerData) Start() error {
+	if err := c.updateStats(); err != nil {
+		return err
+	}
+
 	go c.housekeeping()
 	return nil
 }
@@ -501,12 +505,12 @@ func (c *containerData) housekeepingTick(timer <-chan time.Time, longHousekeepin
 	case <-timer:
 	}
 	start := c.clock.Now()
-	err := c.updateStats()
-	if err != nil {
-		if c.allowErrorLogging() {
-			glog.Warningf("Failed to update stats for container \"%s\": %s", c.info.Name, err)
-		}
-	}
+	// err := c.updateStats()
+	// if err != nil {
+	// 	if c.allowErrorLogging() {
+	// 		glog.Warningf("Failed to update stats for container \"%s\": %s", c.info.Name, err)
+	// 	}
+	// }
 	// Log if housekeeping took too long.
 	duration := c.clock.Since(start)
 	if duration >= longHousekeeping {
diff --git a/manager/manager.go b/manager/manager.go
index 7e665be9..bad085ed 100644
--- a/manager/manager.go
+++ b/manager/manager.go
@@ -370,17 +370,17 @@ func (self *manager) Start() error {
 	glog.V(2).Infof("Recovery completed")
 
 	// Watch for new container.
-	quitWatcher := make(chan error)
-	err = self.watchForNewContainers(quitWatcher)
-	if err != nil {
-		return err
-	}
-	self.quitChannels = append(self.quitChannels, quitWatcher)
+	// quitWatcher := make(chan error)
+	// err = self.watchForNewContainers(quitWatcher)
+	// if err != nil {
+	// 	return err
+	// }
+	// self.quitChannels = append(self.quitChannels, quitWatcher)
 
 	// Look for new containers in the main housekeeping thread.
-	quitGlobalHousekeeping := make(chan error)
-	self.quitChannels = append(self.quitChannels, quitGlobalHousekeeping)
-	go self.globalHousekeeping(quitGlobalHousekeeping)
+	// quitGlobalHousekeeping := make(chan error)
+	// self.quitChannels = append(self.quitChannels, quitGlobalHousekeeping)
+	// go self.globalHousekeeping(quitGlobalHousekeeping)
 
 	return nil
 }
@@ -582,6 +582,10 @@ func (self *manager) getContainer(containerName string) (*containerData, error)
 }
 
 func (self *manager) getSubcontainers(containerName string) map[string]*containerData {
+	if err := self.detectSubcontainers(containerName); err != nil {
+		glog.Errorf("Error detecting containers: %s", err)
+	}
+
 	self.containersLock.RLock()
 	defer self.containersLock.RUnlock()
 	containersMap := make(map[string]*containerData, len(self.containers))
@@ -591,6 +595,10 @@ func (self *manager) getSubcontainers(containerName string) map[string]*containe
 	for i := range self.containers {
 		name := self.containers[i].info.Name
 		if name == containerName || strings.HasPrefix(name, matchedName) {
+			if err := self.containers[i].updateStats(); err != nil {
+				glog.Errorf("Error updating stats for container %s: %s", name, err)
+			}
+
 			containersMap[self.containers[i].info.Name] = self.containers[i]
 		}
 	}

```

TL;DR: only detect new containers and refresh stats on scrape.

This is how it affected CPU usage on one of our machines where `cadvisor` used to take 50% of one core (this is an old Xeon E5-2630 v2 @ 2.60GHz):

![image](https://user-images.githubusercontent.com/89186/42257359-1cc33a44-7f0b-11e8-9b68-74190bbb46cf.png)

That's a nice 5x-6x drop in CPU usage with no effect on what metrics we get.

Flamegraph on a machine with these changes:

![image](https://user-images.githubusercontent.com/89186/42257615-2c033008-7f0c-11e8-9583-5908b379c28f.png)

Flamegraph on a similar machine without these changes:

![image](https://user-images.githubusercontent.com/89186/42257603-17d4a8a0-7f0c-11e8-952f-f82ae1c18b05.png)

**Comment by dashpole** [comment]
Since cAdvisor serves endpoints other than prometheus, and we don't want to make them all update metrics on each query, we can't make the change you are suggesting.

The correct change to make for you is to increase your housekeeping interval, and also increase the window you are taking your rate over.  Try making the window 4x the scrape interval, and making the scrape interval 2x the collection interval.  The ratio of these, rather than the absolute durations is what matters for obtaining an accurate graph.  Tweak them until it is close enough.

**Comment by bobrik** [comment]
> Since cAdvisor serves endpoints other than prometheus, and we don't want to make them all update metrics on each query, we can't make the change you are suggesting.

I'm suggesting to add a mode for people who do not need other endpoints. Prometheus seems to be a de-facto system to manage metrics and it seems useful to support it better with much lower resource requirements.

> The correct change to make for you is to increase your housekeeping interval, and also increase the window you are taking your rate over. Try making the window 4x the scrape interval, and making the scrape interval 2x the collection interval. The ratio of these, rather than the absolute durations is what matters for obtaining an accurate graph. Tweak them until it is close enough.

This is going to reduce precision of my metrics. I don't want that.

**Comment by dqminh** [comment]
@dashpole we already have ondemand housekeeping, wouldnt a reasonable quick workaround/compromise to alleviate CPU issues be that:

- user need to increase the housekeeping interval
- add a plug into prometheus endpoint to start ondemand housekeeping when collecting metrics. ( perhaps with optional maxAge so that if we happen to land closely on a housekeeping tick, dont perform additional work )

**Comment by bobrik** [comment]
Prometheus does support passing arguments, see `params`:

* https://prometheus.io/docs/prometheus/latest/configuration/configuration/#%3Cscrape_config%3E

You can also have a completely different endpoint for ondemand metrics.

**Comment by dashpole** [comment]
In that case, I think it is acceptable to add a MaxAge parameter to the prometheus endpoint to match the json API.  I'm willing to review a PR that accomplishes this.

**Comment by dqminh** [comment]
SGTM, i will take a look at this.
